eloquentjs/4_4_deep_comparison.js

```diff
@@ -0,0 +1,71 @@
+/**
+ * Performs deep comparison of two types and determines equality
+ * @param {*} itemA 
+ * @param {*} itemB 
+ * @returns 
+ */
+function deepEqual(itemA, itemB) {
+    if(itemA === null && itemB === null) {
+        return true;
+    }
+    if (typeof itemA === typeof itemB) {
+        if (typeof itemA === 'object') {
+            if(itemA === null || itemB === null) {
+                return false;
+            }
+
+            let keysA = Object.entries(itemA);
+            let keysB = Object.entries(itemB);
+
+            if (keysA.length !== keysB.length) {
+                return false;
+            }
+
+            if (keysA.length === 0) {
+                throw "objects being compared have no enumerable data types";
+            }
+            
+            for (let i = 0; i < keysA.length; i++) {
+                let keyValuePairA = keysA[i];
+                let keyValuePairB = keysB[i];
+                if (deepEqual(keyValuePairA[0], keyValuePairB[0]) === false || deepEqual(keyValuePairA[1], keyValuePairB[1]) === false) {
+                    return false;
+                }
+            }
+
+            return true;
+        } else {
+            return (itemA === itemB) ? true : false;
+        }
+    }
+    else { 
+        return false;
+    }
+}
+
+let jimFavorites    = {favoriteNumber: 3, favoriteFruit: 'apple', favoriteDog: {name: 'simon', breed: 'border collie'}, favoriteColors: ['red', 'blue']};
```

Comment:
You're missing some important test cases. The linked article may give you enough info to come up with them. See if you can!